### PR TITLE
Add pool unit tests

### PR DESCRIPTION
**Pull request by amityadav0**

### CODE REVIEW

File: nearswap/src/pool.rs [hunk 1]
```diff
@@ -170,3 +169,108 @@ impl Pool {
...
+
+        assert!(pool.ynear == 100, "liquidity added is incorrect");
+        assert!(pool.tokens == 200, "liquidity added is incorrect");
+        assert!(pool.total_shares == 100, "liquidity added is incorrect");
+
+        let (near_before, tokens_before, shares_before) = (pool.ynear, pool.tokens, pool.total_shares);
+        let (near, tokens, shares) = expected_added_liquidity(200, 400, &pool);
+        
+        // add liquidity again
+        pool.add_liquidity(&caller, 200, 400, 0);
```
Comment (robert-zaremba) [inline nearswap/src/pool.rs:250]: let's validate return value here

File: nearswap/src/pool.rs [hunk 2]
```diff
@@ -170,3 +169,108 @@ impl Pool {
...
+
+        assert!(near_before + near == pool.ynear, "liquidity added is incorrect");
+        assert!(tokens_before + tokens == pool.tokens, "liquidity added is incorrect");
+        assert!(shares_before + shares == pool.total_shares, "liquidity added is incorrect");
```
Comment (robert-zaremba) [inline nearswap/src/pool.rs:254]: let's add one more step and try to add liquidity with different ratio, eg:  `(100, 100)` and validate return value.

File: nearswap/src/pool.rs [hunk 3]
```diff
@@ -170,3 +169,108 @@ impl Pool {
...
+        let (near_before, tokens_before, shares_before) = (pool.ynear, pool.tokens, pool.total_shares);
+        let (expected_near, expected_tokens) = expected_withdraw(50, &pool);
+
+        pool.withdraw_liquidity(&caller, 0, 0, 50);
+
+        // withdraw shares
+        assert!(pool.ynear == near_before - expected_near, "liquidity removed is incorrect");
+        assert!(pool.tokens == tokens_before - expected_tokens, "liquidity removed is incorrect");
+        assert!(pool.total_shares == shares_before - 50, "liquidity removed is incorrect");
+    }
```
Comment (robert-zaremba) [inline nearswap/src/pool.rs:275]: let's add a failing scenario (eg, when we a user who doesn't have LP tries to withdraw and when a user with LP tries to withdraw more than he has).

**Review by robert-zaremba — COMMENTED** [review]
Let's add more tests

File: nearswap/src/pool.rs (hunk 1 above)
Comment (amityadav0) [reply nearswap/src/pool.rs:250]: Ok

File: nearswap/src/pool.rs (hunk 2 above)
Comment (amityadav0) [reply nearswap/src/pool.rs:254]: Ok

File: nearswap/src/pool.rs (hunk 3 above)
Comment (amityadav0) [reply nearswap/src/pool.rs:275]: Ok

File: nearswap/src/pool.rs [hunk 4]
```diff
@@ -170,3 +169,178 @@ impl Pool {
...
+
+    #[test]
+    #[should_panic(expected = r#"attempt to subtract with overflow"#)]
+    fn withdraw_liquidity_pool_fail_scenario_1() {
```
Comment (robert-zaremba) [inline nearswap/src/pool.rs:320]: please add a comment describing the scenario. For the test case below - please add the description as well.

File: nearswap/src/pool.rs [hunk 5]
```diff
@@ -170,3 +169,178 @@ impl Pool {
...
+
+        assert!(near_added == expected_near, "liquidity added is incorrect");
+        assert!(tokens_added == expected_tokens, "liquidity added is incorrect");
+        assert!(shares_minted == expected_shares, "liquidity added is incorrect");
+
+        let (expected_near2, expected_tokens2, expected_shares2) = expected_added_liquidity(100, 100, &pool);
+        // add liquidity again with ratio 1:1(100:100)
+        let (near_added2, tokens_added2, shares_minted2) = pool.add_liquidity(&caller, 100, 100, 0);
+
+        assert!(near_added2 == expected_near2, "liquidity added is incorrect");
```
Comment (robert-zaremba) [inline nearswap/src/pool.rs:259]:
```suggestion
        assert!(near_added2 == expected_near2, "liquidity added is incorrect");
        assert!(near_added2 == 100, "liquidity added is incorrect");
```

**Review by robert-zaremba — APPROVED** [review]
left 3 comments.

File: nearswap/src/pool.rs (hunk 4 above)
Comment (amityadav0) [reply nearswap/src/pool.rs:320]: ok

File: nearswap/src/pool.rs (hunk 5 above)
Comment (amityadav0) [reply nearswap/src/pool.rs:259]: shares here would be 51 because pool ratio is (1:2) so user gives 100 tokens therfore near is adjusted to half.

Comment (robert-zaremba) [reply nearswap/src/pool.rs:259]: right, then let's put 51 here.

**Pull request merged**